### DISCUSSION TITLE
A way to safely fill a `reinforced air canister` completely

### DISCUSSION THREAD

**Discussion by Quezler** [TeamPneumatic/pnc-repressurized#994] [Ideas]
The `reinforced air canister` needs 20.0 bar to be considered full, to my knowledge there is currently no way to safely fill it completely since the `charging station` and `charging module` need to be of a pressure above it in order to fill them, which puts them in the danger zone.

Even without a security module in the pump (so you can get the pressure in the 20.0 - 20.4 range, instead of just 19.8) by using external redstone you are in the danger zone, since[ no part of the red is safe](https://github.com/TeamPneumatic/pnc-repressurized/blob/872003157d152f48d542d92be0dffd12ddd9b6db/src/main/java/me/desht/pneumaticcraft/common/capabilities/MachineAirHandler.java#L186), effectively making the danger pressure identical to critical, the only difference being no **instant** explosion.

It would be nice if the armor upgrade could be installed in some machines (only caring about the compressor and charging [station) where each upgrade would add e.g. a 0.1 bar safe zone for that machine to be in the red, for example:
[`if ((getAir() - <armor upgrades> * 0.1) > prevAir && p > dangerPressure) {`](https://github.com/TeamPneumatic/pnc-repressurized/blob/872003157d152f48d542d92be0dffd12ddd9b6db/src/main/java/me/desht/pneumaticcraft/common/capabilities/MachineAirHandler.java#L182)

This would allow players to reliably fill reinforced air canisters completely without having to worry about explosions.

Naturally since tubes can't have upgrades installed you'd have to put the compressor & charging station back-to-back or risk blowing the tube, so it might not fit entirely within the design philosophy of pneumaticcraft. 

Perhaps the cleanest solution would be to simply have an universal safe (read: actual danger) margin of 0.25 bar on the normal tier and 0.5 bar on all the advanced ones (or perhaps more, or less, i'm not the best balancer in the world)

And why? well i'm ocd, and seriously bothered by the fact that i can't safely get the reinforced air canisters above 19.9 safely so the durability bar disappears and all the nbts match so they are easier to deal with in other mods and aid in the filtering of items.

Feel free to completely disregard this idea, writing it down is just my way of getting it out of my head (of course i do hope that there will be a way to safely fill them completely without any risk of explosions, e.g. a sort of hydraulic ram where some pressure could be sacrificed in order to stuff the last few digits into a can somehow)

Thanks for reading, here is a potato: 🥔

**Comment by TeamSpen210** [comment]
Maybe just change the canister to stop at 19.9, with a full bar. Or have the charging station special case a little and fill anyway?
